diff --git a/addition/stcore_addition/S095_mangos_vip_commands.sql b/addition/stcore_addition/S095_mangos_vip_commands.sql
index 04e21fa..263b1cc 100644
--- a/addition/stcore_addition/S095_mangos_vip_commands.sql
+++ b/addition/stcore_addition/S095_mangos_vip_commands.sql
@@ -6,3 +6,7 @@ REPLACE INTO `command` (`name`,`security`,`help`) VALUES ('vip repair', 1, 'Synt
 REPLACE INTO `command` (`name`,`security`,`help`) VALUES ('vip auction', 1, 'Syntax: .vip auction - opens auctionhouse window to character.');
 REPLACE INTO `command` (`name`,`security`,`help`) VALUES ('vip resettalents', 1, 'Syntax: .vip resettalents - reset talents to your character.');
 REPLACE INTO `command` (`name`,`security`,`help`) VALUES ('vip whispers', 1, 'Syntax: .vip whispers - Enable/disable accepting whispers by VIP accounts from players.');
+REPLACE INTO `command` (`name`,`security`,`help`) VALUES ('vip taxi', 1, 'Syntax: .vip taxi - reveals all taxipaths to owner');
+
+INSERT INTO `mangos_string` VALUES (12000, 'You can\'t use command while being dead/stealthed or being in arena/bg orbeing in combat.', NULL, NULL, NULL, NULL, NULL, NULL, NULL, 'Вы не можете использовать комманду находясь на арене/БГ, в комбате/стелсе или будучи мертвым.');
+INSERT INTO `mangos_string` VALUES (12001, 'You can\'t use command while being dead/stealthed or being in arena/bg orbeing in combat.', NULL, NULL, NULL, NULL, NULL, NULL, NULL, 'Вы не можете использовать комманду находясь на арене/БГ/подземелье, в комбате/стелсе или будучи мертвым.')
diff --git a/src/game/Chat.cpp b/src/game/Chat.cpp
index 1671421..bf756d1 100644
--- a/src/game/Chat.cpp
+++ b/src/game/Chat.cpp
@@ -660,6 +660,20 @@ ChatCommand * ChatHandler::getCommandTable()
         { NULL,             0,                  false, NULL,                                           "", NULL }
     };
     
+    static ChatCommand vipCommandTable[] =
+    {
+        { "debuff",         SEC_ADMINISTRATOR,  false, &ChatHandler::HandleVipDebuffCommand,           "", NULL },
+        { "map",            SEC_ADMINISTRATOR,  false, &ChatHandler::HandleVipMapCommand,              "", NULL },
+        { "bank",           SEC_ADMINISTRATOR,  false, &ChatHandler::HandleVipBankCommand,             "", NULL },
+        { "repair",         SEC_ADMINISTRATOR,  false, &ChatHandler::HandleVipRepairCommand,           "", NULL },
+        { "auction",        SEC_ADMINISTRATOR,  false, &ChatHandler::HandleVipAuctionCommand,          "", NULL },
+        { "resettalents",   SEC_ADMINISTRATOR,  false, &ChatHandler::HandleVipResetTalentsCommand,     "", NULL },
+        { "whispers",       SEC_ADMINISTRATOR,  false, &ChatHandler::HandleVipWhispersCommand,         "", NULL },
+        { "taxi",           SEC_ADMINISTRATOR,  false, &ChatHandler::HandleVipTaxiCommand,             "", NULL },
+        
+        { NULL,             0,                  false, NULL,                                           "", NULL }
+    };
+    
     static ChatCommand chatspyCommandTable[] =
     {
         { "set",            SEC_ADMINISTRATOR,  false, &ChatHandler::HandleChatSpySetCommand,          "", NULL },
@@ -766,6 +780,7 @@ ChatCommand * ChatHandler::getCommandTable()
         { "unfreeze",       SEC_ADMINISTRATOR,     true, &ChatHandler::HandleUnFreezeCommand,            "", NULL },
         { "listfreeze",     SEC_ADMINISTRATOR,     true, &ChatHandler::HandleListFreezeCommand,          "", NULL },
         { "chatspy",        SEC_ADMINISTRATOR,     true, NULL,                                           "", chatspyCommandTable },
+        { "vip",            SEC_ADMINISTRATOR,     true, NULL,                                           "", vipCommandTable },
 
         { NULL,             0,                  false, NULL,                                           "", NULL }
     };
diff --git a/src/game/Chat.h b/src/game/Chat.h
index ccd6e42..23b7001 100644
--- a/src/game/Chat.h
+++ b/src/game/Chat.h
@@ -574,6 +574,14 @@ class ChatHandler
         bool HandleFreezeCommand(char *args);
         bool HandleUnFreezeCommand(char *args);
         bool HandleListFreezeCommand(char* args);
+        bool HandleVipDebuffCommand(char* args);
+        bool HandleVipMapCommand(char* args);
+        bool HandleVipBankCommand(char* args);
+        bool HandleVipRepairCommand(char* args);
+        bool HandleVipAuctionCommand(char* args);
+        bool HandleVipResetTalentsCommand(char* args);
+        bool HandleVipWhispersCommand(char* args);
+        bool HandleVipTaxiCommand(char* args);
 
         bool HandleAHBotOptionsCommand(char* args);
 
diff --git a/src/game/Language.h b/src/game/Language.h
index 450a84a..1d99905 100644
--- a/src/game/Language.h
+++ b/src/game/Language.h
@@ -934,6 +934,10 @@ enum MangosStrings
     LANG_CHATSPY_LISTOFSPYS             = 11606,
     LANG_CHATSPY_TOTAL                  = 11607,
     LANG_CHATSPY_ONESPYSANOTHER         = 11608,
+    
+    //VIP
+    VIP_CANT_DO                         = 12000,
+    VIP_CANT_DO_2                       = 12001,
 
     // Use for custom patches             11000-11999
 
diff --git a/src/game/Level3.cpp b/src/game/Level3.cpp
index 57b986d..08e1328 100644
--- a/src/game/Level3.cpp
+++ b/src/game/Level3.cpp
@@ -7342,6 +7342,171 @@ bool ChatHandler::HandleListFreezeCommand(char* args)
     return true;
 }
 
+bool ChatHandler::HandleVipDebuffCommand(char* /*args*/)
+{   
+    Player *chr = m_session->GetPlayer();
+
+    //Different Checks
+    if( chr->isInCombat() || chr->InBattleGround() || chr->HasStealthAura() || chr->HasFlag(UNIT_FIELD_FLAGS_2, UNIT_FLAG2_FEIGN_DEATH) || chr->isDead())
+    {
+        SendSysMessage(VIP_CANT_DO);
+        SetSentErrorMessage(true);
+        return false;
+    }
+    
+    m_session->GetPlayer()->RemoveAurasDueToSpell(SPELL_ID_PASSIVE_RESURRECTION_SICKNESS);
+    m_session->GetPlayer()->RemoveAurasDueToSpell(26013);
+
+    return true;
+}
+
+bool ChatHandler::HandleVipMapCommand(char* /*args*/)
+{
+    Player *chr = m_session->GetPlayer();
+
+    //Different Checks
+    if( chr->isInCombat() || chr->InBattleGround() || chr->HasStealthAura() || chr->HasFlag(UNIT_FIELD_FLAGS_2, UNIT_FLAG2_FEIGN_DEATH) || chr->isDead())
+    {
+        SendSysMessage(VIP_CANT_DO);
+        SetSentErrorMessage(true);
+        return false;
+    }
+
+    PSendSysMessage(LANG_YOU_SET_EXPLORE_ALL, GetNameLink(m_session->GetPlayer()).c_str());
+    for (uint8 i=0; i<PLAYER_EXPLORED_ZONES_SIZE; ++i)
+    {
+        m_session->GetPlayer()->SetFlag(PLAYER_EXPLORED_ZONES_1+i,0xFFFFFFFF); 
+    }
+
+    return true;
+}
+
+bool ChatHandler::HandleVipBankCommand(char* /*args*/)
+{
+    Player *chr = m_session->GetPlayer();
+
+    //Different Checks
+    if( chr->isInCombat() || chr->InBattleGround() || chr->HasStealthAura() || chr->HasFlag(UNIT_FIELD_FLAGS_2, UNIT_FLAG2_FEIGN_DEATH) || chr->isDead())
+    {
+        SendSysMessage(VIP_CANT_DO);
+        SetSentErrorMessage(true);
+        return false;
+    }
+
+    m_session->SendShowBank(m_session->GetPlayer()->GetObjectGuid());
+
+    return true;
+}
+
+bool ChatHandler::HandleVipRepairCommand(char* /*args*/)
+{
+    Player *chr = m_session->GetPlayer();
+
+    //Different Checks
+    if( chr->isInCombat() || chr->InBattleGround() || chr->HasStealthAura() || chr->HasFlag(UNIT_FIELD_FLAGS_2, UNIT_FLAG2_FEIGN_DEATH) || chr->isDead())
+    {
+        SendSysMessage(VIP_CANT_DO);
+        SetSentErrorMessage(true);
+        return false;
+    }
+    
+    // Repair items
+    m_session->GetPlayer()->DurabilityRepairAll(false, 0, false);
+
+    PSendSysMessage(LANG_YOUR_ITEMS_REPAIRED, GetNameLink(m_session->GetPlayer()).c_str());
+    return true;
+}
+
+bool ChatHandler::HandleVipAuctionCommand(char* /*args*/)
+{
+    Player *chr = m_session->GetPlayer();
+
+    //Different Checks
+    if( chr->isInCombat() || chr->InBattleGround() || chr->HasStealthAura() || chr->HasFlag(UNIT_FIELD_FLAGS_2, UNIT_FLAG2_FEIGN_DEATH) || chr->isDead())
+    {
+        SendSysMessage(VIP_CANT_DO);
+        SetSentErrorMessage(true);
+        return false;
+    }
+
+    m_session->GetPlayer()->SetAuctionAccessMode(0);
+    m_session->SendAuctionHello(m_session->GetPlayer());
+
+    return true;
+}
+
+bool ChatHandler::HandleVipResetTalentsCommand(char* /*args*/)
+{
+    Player *chr = m_session->GetPlayer();
+
+    //Different Checks
+    if( chr->isInCombat() || chr->InBattleGround() || chr->HasStealthAura() || chr->HasFlag(UNIT_FIELD_FLAGS_2, UNIT_FLAG2_FEIGN_DEATH) || chr->isDead())
+    {
+        SendSysMessage(VIP_CANT_DO);
+        SetSentErrorMessage(true);
+        return false;
+    }
+
+    // Reset Talents
+    m_session->GetPlayer()->resetTalents(true);
+    m_session->GetPlayer()->SendTalentsInfoData(false);
+
+    PSendSysMessage(LANG_RESET_TALENTS_ONLINE, GetNameLink(m_session->GetPlayer()).c_str());
+    return true;
+}
+
+
+bool ChatHandler::HandleVipWhispersCommand(char* args)
+{
+    if(!*args)
+    {
+        PSendSysMessage(LANG_COMMAND_WHISPERACCEPTING, m_session->GetPlayer()->isAcceptWhispers() ?  GetMangosString(LANG_ON) : GetMangosString(LANG_OFF));
+        return true;
+    }
+
+    bool value;
+    if (!ExtractOnOff(&args, value))
+    {
+        SendSysMessage(LANG_USE_BOL);
+        SetSentErrorMessage(true);
+        return false;
+    }
+
+    // whisper on
+    if (value)
+    {
+        m_session->GetPlayer()->SetAcceptWhispers(true);
+        SendSysMessage(LANG_COMMAND_WHISPERON);
+    }
+    // whisper off
+    else
+    {
+        m_session->GetPlayer()->SetAcceptWhispers(false);
+        SendSysMessage(LANG_COMMAND_WHISPEROFF);
+    }
+
+    return true;
+}
+
+bool ChatHandler::HandleVipTaxiCommand(char* /*args*/)
+{
+    Player *chr = m_session->GetPlayer();
+
+    //Different Checks
+    if( chr->isInCombat() || chr->InBattleGround() || chr->HasStealthAura() || chr->HasFlag(UNIT_FIELD_FLAGS_2, UNIT_FLAG2_FEIGN_DEATH) || chr->isDead())
+    {
+        SendSysMessage(VIP_CANT_DO_2);
+        SetSentErrorMessage(true);
+        return false;
+    }
+    
+    chr->SetTaxiCheater(true);
+    PSendSysMessage(LANG_YOU_GIVE_TAXIS, GetNameLink(chr).c_str());
+    if (needReportToTarget(chr))
+        ChatHandler(chr).PSendSysMessage(LANG_YOURS_TAXIS_ADDED, GetNameLink().c_str());
+    return true;
+}
+
 bool ChatHandler::HandleModifyGenderCommand(char *args)
 {
     if (!*args)
